# Tarea_2 - Parte 2

In [1]:
import numpy as np

## Ejercicio 1

En ML usamos la entropía cruzada como una forma de medir que tan bueno es un modelo de variables discretas a través de comparar la distribución de probabilidad que el modelo produce o predice, vs la distribución de probabilidad real dada por los datos de entrenamiento.

Podemos ver el siguiente ejemplo que define la forma en que se calcula la entropía cruzada y nos muestra un caso específico:

<img src="https://image.slidesharecdn.com/publishintroductiontodeeplearninginpythonandmatlab1-160502102437/95/introduction-to-deep-learning-in-python-and-matlab-54-638.jpg?cb=1462185644">

En este caso interpretamos así: El problema consiste en un modelo o algoritmo de ML que debe producir un vector de 3 elementos indicando la probabilidad de que ciertos datos X pertenezcan a una de 3 categorías.

* El modelo de ML produce un vector que indica que estima un 70% de probabilidad de que se trate de la categoría 0, 20% de probabilidad de que se trate de la categoría 1 y 10% de que se trate de la categoría 2.
* Los datos reales nos dicen que se trataba de un caso donde con total certeza se sabe que se trata de la categoría 0
* La entropía cruzada(a calcular en el ejercicio) nos indica que tan buena es la estimación del modelo, una EC de 0 es un modelo perfecto(en este caso un modelo que predice 100% de prob para la clase 0)

**Nota** 
* Aun que para calcular la entropía usamos logaritmos en base 2, en ML para calcular la entropía cruzada se usa logaritmo natural ya que con este se cumple el proposito **estimar que tanto se alejan las predicciones del modelo de ML de los datos reales** y es comun]mente mas rápido de calcular en la computadora.
* Ya que estamos trabajando con vectores que representan distribuciones de probabilidad , podemos toparnos con lo que se conoce como : **sparse vectors**(vectores donde la mayoría de elementos son 0), esto puede producir problemas ya que le logaritmo de 0 no esta definido, tu solución debe tomar en cuenta esto y evitar que devuelva "nan"

In [47]:
def cross_entropy(Y,Y_hat):
    ## tu codigo aqui (~ 1 lineas de codigo)
    booleanos = Y_hat > 0
    return -1*np.sum(Y[booleanos]*np.log(Y_hat[booleanos]))  

y  = np.array([1.0,0,0])
y_hat = np.array([0.7,0.2,0.1])

cross_entropy(y,y_hat)

0.35667494393873245


# Magnitud de un Vector(norma o módulo)
 

Matemáticamente la magnitud de un vector(también conocida como norma vectorial)  nos indica  el tamaño de este , y nos sirve para tener una noción de la distancia desde un punto de referencia(origen) hasta el punto representado por el vector.


## Ejercicio 2
Crear una función que reciba como parámetro un vector x y calcule su magnitud o norma(euclidiana o L2) ,luego usarla para evaluar 2 vectores que representan los errores generados por 2 modelos de machine learning y concluir cual de los 2 modelos es mejor.

In [5]:
def magnitud(x):
    #Tu codigo aqui (~ 1 linea de codigo)
    #La norma euclidiana no solo se limita a vectores de dos o tres posiciones
    return np.sqrt(np.sum(np.power(x,2)))

errores_modelo1 = np.array([1,2,1,2])
errores_modelo2 = np.array([0,1,1,2])

print(magnitud(errores_modelo1))
print(magnitud(errores_modelo2))

3.1622776601683795
2.449489742783178


### Respuesta:
El mejor modelo es el segundo por tener un error menor.

## Ejercicio 3
Usando la función del ejercicio anterior, crea otra función normalizar(x) que reciba de parámetro un vector x aplique normalización sobre este, el resultado debe ser un nuevo vector del tamaño de x cuya magnitud es igual a 1.0

In [6]:
def normalizar(x):
    ## tu codigo aqui (~ 3 linea de codigo)
    return x/np.sqrt(np.sum(np.power(x[np.arange(0, x.size)],2)))

print(magnitud(normalizar(errores_modelo1)))
print(normalizar(errores_modelo2))

1.0
[0.         0.40824829 0.40824829 0.81649658]


# Producto Punto(escalar,interno,interior) y Ortogonalidad



## Ejercicio 4
### Aplicado en DS

Se tiene una red neuronal sencilla(y simplificada) como la de la siguiente imagen:
<img src="https://www.oreilly.com/library/view/practical-convolutional-neural/9781788392303/assets/246151fb-7893-448d-b9bb-7a87b387a24b.png">

Donde:
* INPUT LAYER: un vector X de tamaño = 2 que representa los datos de entrada
* HIDDEN_LAYER :capa oculta con 2 neuronas definidas por los vectores:
    * HL1 = [0.25,0.37]
    * HL2 = [-8,14]
* OUTPUT_LAYER = capa de salida definida por el vector [4,9]

Crear una funcion `neural_network(X)` para calcular:
* Calcule la salida de cada neurona en la capa intermedia aplicada a la capa de entrada.
* Use el resultado del paso anterior como entrada para la neurona en la capa de salida

Asumiendo que cada neurona identifica la similitud entre su entrada y la caracteística que representa concluir:
* Para cada vector de entrada Xi , cual neurona intermedia busca la característica que mas se parece a X.
* Cual vector de entrada Xi produce una activación alta(salida alta) en la capa de salida.

In [20]:
# Aplicar la red neuronal sobre los siguientes datos X

X1 = np.array([0.50,0.72])
X2 = np.array([-4,7])
X3 = np.zeros_like(X2)
X4 = np.ones_like(X1)
X5 = np.random.randn(X1.shape[0])

In [22]:
## tu codigo aqui:

def neural_network(x):
    HL1 = np.array([0.25,0.37])
    HL2 = np.array([-8, 14])
    OL = np.array([4,9])
    #r1 es el resultado de cada neurona en la capa intermedia
    r1 = np.hstack((np.dot(x, HL1), np.dot(x, HL2)))
    return np.dot(r1, OL)
    
print(neural_network(X1))
print(neural_network(X2))
print(neural_network(X3))
print(neural_network(X4))
print(neural_network(X5))

56.2856
1176.36
0.0
56.48
-177.72575578863254


### Respuesta:

***1. Para cada vector de entrada Xi , cual neurona intermedia busca la característica que mas se parece a X.***

Para X1 la neurona que más se parece es HL2 [0.3914, 6.08]

Para X2 la neurona que más se parece es HL2 [1.59, 130.]

Para X3 ambas neuronas se parecen [0., 0.]

Para X4 la neurona que más se parece es HL2 [0.62, 6.]

Para X5 la neurona que más se parece es HL2 [0.04348972, -19.76663496]

***2. Cual vector de entrada Xi produce una activación alta (salida alta) en la capa de salida.***

X2 produce la activación más alta con una salida igual a 1176.36

## Ejercicio 5 (aplicado en DS)
La correlación cruzada es una medida de similitud entre 2 funciones como resultado de "desplazar" una sobre la otra, comunmente es usada para encontrar características relevantes en una función desconocida o no controlada.

En procesamiento de señales por ejemplo es usada para buscar y/o filtrar en una señal que varia en el tiempo cierta caracaterística de interés. Aveces es llamada también "sliding dot product" consiste en aplicar en cada punto de una función F el producto punto con cierta función G(comunmente mas corta) y luego "deslizar" G a un nuevo punto de la función, el resultado es una nueva función H que se interpreta como :**cuanto se parece** en cada punto la función F a la característica G. 

<img src="https://i.makeagif.com/media/11-25-2015/LZ9Ufj.gif"  height="250" width="300">

En este ejercicio usamos correlación cruzada para calcular las medias moviles promediando 3 puntos que ya vimos en otro ejemplo:

El primer paso es definir la función G que define el "filtro" a aplicar, para este caso consiste simplemente en un vector con 3 elementos donde cada elemento corresponde a 1/3, luego debemos aplicar el producto punto sobre cada punto de la función o datos originales(en este ejemplo llamados x) a traves aplicar el producto punto en cada elemento "corriendolo" de uno en uno

In [18]:
ruido = 0.1*np.random.randn(15) #el ruido comunmente se debe a aleatoriedad o captura no exacta de info.
x = np.linspace(0,2*np.pi,15) + ruido
x = np.sin(x)

filtro = np.array([1/3,1/3,1/3])

## tu codigo aqui (~ 3 linea de codigo)

xs = x.copy()

for z in range(1, x.size-1):
    xs[z] = np.dot(np.hstack((x[z-1], x[z], x[z+1])), filtro)
    
print(x)
print(xs)

[-0.14242698  0.54134193  0.75342371  0.90699164  0.99225583  0.62658899
  0.55219864 -0.06874572 -0.55164921 -0.69919276 -0.99239733 -0.95511272
 -0.71710508 -0.40552614  0.00099438]
[-0.14242698  0.38411289  0.73391909  0.88422373  0.84194548  0.72368115
  0.37001397 -0.0227321  -0.43986257 -0.74774643 -0.88223427 -0.88820504
 -0.69258131 -0.37387895  0.00099438]


# Operadores logicos en vectores

## Ejercicio 6
Dado el vector x, usar operadores lógicos sobre vectores y acceso a  elementos usando vectores booleanos , escribir un programa que calcule eun nuevo vector z conteniendo el valor absoluto de el vector x.

**Nota** No se puede usar np.abs() ni ciclos

**Tip** usar un vector booleano para saber que elementos son negativos y deben ser multiplicados por -1

In [9]:
x = np.linspace(-2,2,9)
## tu codigo aqui (~ 4 linea de codigo)
x[x<0] *= -1
print(x)

[2.  1.5 1.  0.5 0.  0.5 1.  1.5 2. ]




## Ejercicio 7
Implementar la función:
$$h(x) = \begin{cases}0 & x< 0\\1 & 0<=x

$$h(x) = \begin{cases}0 & x< 0\\1 & 0<=x <=1\\0 & x> 1\end{cases}$$

Esta debe funcionar para vectores de cualquier tamaño x:

```
def h(x):
    
    ...
    
```
**nota** debe ser implementada sin ciclos o ifs 

In [41]:
def h(x):
    ## tu codigo aqui (~ 3 linea de codigo
    x[(x >= 0) & (x <= 1)] = 1
    x[x < 0] = 0
    x[x > 1] = 0
    return x

x = np.array([0.1,-2,0.5,5])

print(h(x))

[1. 0. 1. 0.]


Para casos donde se requiere comportamiento parecido a este pero con mas condiciones, NumPy provee la función **np.select** , la descripción de esta función casi siempre es mas complicada de lo que debería y dificulta entenderla, vamos a buscar entenderla bajo un ejemplo.

Básicamente select se basa en :
* una lista de condiciones A
* una lista de valores a tomar según estas condiciones B
* un valor de  resultado default cuando ninguna de las condiciones en la  lista se cumple.
Estas 2 listas deben ser del mismo tamañaño.

# Select y performance

Usando select implementar una función k(x) con la siguiente definición:
$$k(x) = \begin{cases}-x & x< 0\\x^{3}  &0<=x<1\\x^{2}  &1<=x<2\\4  &otherwise\end{cases}$$



Aun que select es muy conveniente y útil , no es la opción mas eficiente esto debido a que evalua todas las condiciones y todos los resultados, en programación en general buscamos escribir los programas de la manera mas eficiente posible y evitar calculos innecesarios, esto se vuelve vital en ciencia de datos cuando procesamos grandes volúmenes de información, por eso NumPy nos provee la alternativa a select llamada **piecewise** que funciona de manera similar .

Este funciona similar a select pero en vez de calcular todos los posibles resultados, calcula solo aquellos para los que la condición es True e ignora los False,sintácticamente piecewise requiere que los "resultados" sesan calcuados usando una lista de funciones por lo cual todos los resultados deben estar contenidos en una función  y el objeto función ser enviado a piecewise (si la función solo se utiliza una vez para este propósito, se puede usar funciones anónimas o lambda).

## Ejercicio 8

Investigar piecewise:
*  Usarlo para implementar la función anterior de manera eficiente
*  Usar piscewise para implementar la función:

$$m(x) = \begin{cases}e^{2x} & x< 0\\1  &0<=x<=1\\e^{1-x}  &x<=1\\\end{cases}$$

Nuevamente, sin utilizar ciclos ni ifs

In [15]:
def m(x):
    return np.piecewise(x,[x<0,(x>= 0) & (x<1), x>=1],[lambda x: np.exp(2*x),1,lambda x: np.exp(1-x)])

x = np.array([0.1,-2,0.5,5])
print(m(x))

[1.         0.01831564 1.         0.01831564]
